parsers/regex/regex_test.go

```diff
@@ -0,0 +1,168 @@
+		input:     `[some value] unbracketed`,
+		expected: map[string]interface{}{
+			"BracketedField":   "some value",
+			"UnbracketedField": "unbracketed",
+		},
+	},
+	{
+		// Nested regex grouping
+		lineRegex: `(?P<outer>[^ ]* (?P<inner1>[^ ]*) (?P<inner2>[^ ]*))`,
+		input:     `foo bar baz`,
+		expected: map[string]interface{}{
+			"outer":  "foo bar baz",
+			"inner1": "bar",
+			"inner2": "baz",
+		},
+	},
+	{
+		// Sample nginx log line
+		lineRegex: `(?P<time>\d{4}/\d{2}/\d{2} \d{2}:\d{2}:\d{2}) \[(?P<status>.*)\].* request: "(?P<request>[^"]*)"`,
+		input:     `2017/11/07 22:59:46 [error] 5812#0: *777536449 connect() failed (111: Connection refused) while connecting to upstream, client: 127.0.0.1, server: localhost, request: "GET /isbns HTTP/1.1", upstream: "http://127.0.0.1:8080/isbns", host: "localhost"`,
+		expected: map[string]interface{}{
+			"time":    "2017/11/07 22:59:46",
+			"status":  "error",
+			"request": "GET /isbns HTTP/1.1",
+		},
+	},
+}
+
+func TestParseLine(t *testing.T) {
+	p := &Parser{
+		conf: Options{
+			NumParsers: 5,
+		},
+		lineParser: &RegexLineParser{regexp.MustCompile(``)},
+	}
+	for _, tlm := range tlms {
+		p.lineParser = &RegexLineParser{regexp.MustCompile(tlm.lineRegex)}
+		resp, err := p.lineParser.ParseLine(tlm.input)
+		t.Logf("%+v", resp)
+		if err != nil {
+			t.Error("p.ParseLine unexpectedly returned error ", err)
+		}
+		if !reflect.DeepEqual(resp, tlm.expected) {
+			t.Errorf("response %+v didn't match expected %+v", resp, tlm.expected)
+		}
+	}
+}
+
+type testLineMaps struct {
+	line        string
+	trimmedLine string
+	resp        map[string]interface{}
+	typedResp   map[string]interface{}
+	ev          event.Event
+}
+
+type FakeLineParser struct {
+	tlm []testLineMaps
+}
+
+func (f *FakeLineParser) ParseLine(line string) (map[string]interface{}, error) {
+	for _, pair := range f.tlm {
+		if pair.trimmedLine == line {
+			return pair.resp, nil
+		}
+	}
+	return nil, errors.New("failed to parse line")
+}
+
+// Test event emitted from ProcessLines
+func TestProcessLines(t *testing.T) {
+	t1, _ := time.ParseInLocation(commonLogFormatTimeLayout, "08/Oct/2015:00:26:26 -0000", time.UTC)
+	preReg := &parsers.ExtRegexp{regexp.MustCompile("^.*:..:.. (?P<pre_hostname>[a-zA-Z-.]+): ")}
+	tlm := []testLineMaps{
+		{
+			line: "https - 10.252.4.24 - - [08/Oct/2015:00:26:26 +0000] 200 174 0.099",
+			resp: map[string]interface{}{
+				"http_x_forwarded_proto": "https",
+				"remote_addr":            "10.252.4.24",
+				"time_local":             "08/Oct/2015:00:26:26 +0000",
+			},
+			ev: event.Event{
+				Timestamp: t1,
+				Data: map[string]interface{}{
+					"http_x_forwarded_proto": "https",
+					"remote_addr":            "10.252.4.24",
+				},
+			},
+		},
+	}
+	lineRegex, err := regexp.Compile(`(?P<http_x_forwarded_proto>\w+) - (?P<remote_addr>\d{1,4}\.\d{1,4}\.\d{1,4}\.\d{1,4}) - - \[(?P<local_time>\d{2}\/[A-Za-z]+\/\d{4}:\d{2}:\d{2}:\d{2}.*)\]`)
```

Comment:
I'm confused - how is this test passing? The named group is `local_time` but the `resp` map above lists `time_local` as the expected key.
oh! `pair.resp` isn't actually used anywhere - the `resp` down below is the parsed line, not the map from the test definition. 
Should we drop `resp` from the test setup or add a check down below to check it? I lean towards dropping it from the test definition since this test is using the `ProcessLines` interface rather than the `LineParser` so won't have access to the map, only the event.